packages/paymaster/src/BiconomyPaymaster.ts

```diff
@@ -1,49 +1,241 @@
-import { sendRequest, HttpMethod } from '@biconomy/common'
+import { Logger, sendRequest, HttpMethod } from '@biconomy/common'
 import { resolveProperties } from '@ethersproject/properties'
-import { UserOperation } from '@biconomy/core-types'
-import { IPaymaster } from 'interfaces/IPaymaster'
-import { PaymasterAndDataResponse, PaymasterConfig } from './utils/Types'
+import { UserOperation, Transaction } from '@biconomy/core-types'
+import { Provider } from '@ethersproject/abstract-provider'
+import {
+  PaymasterFeeQuote,
+  PaymasterConfig,
+  BiconomyTokenPaymasterFeeQuoteResponse,
+  FeeQuotesOrDataDto,
+  SponsorUserOperationDto,
+  PaymasterServiceSuccessResponse,
+  BiconomyTokenPaymasterRequest
+} from './utils/Types'
+import { BigNumberish, BigNumber, ethers } from 'ethers'
+// TODO
+import { ERC20_ABI } from './constants' // temporary
+import { IHybridPaymaster } from './interfaces/IHybridPaymaster'
 
-export class BiconomyPaymaster implements IPaymaster {
+// Hybrid - Generic Gas abstraction paymaster
+/**
+ *
+ */
+export class BiconomyPaymaster implements IHybridPaymaster {
   constructor(readonly paymasterConfig: PaymasterConfig) {}
+
+  // TODO // WIP
+  // Note: maybe rename to createTokenApprovalTransaction
+  async createTokenApprovalRequest(
+    tokenPaymasterRequest: BiconomyTokenPaymasterRequest,
+    provider: Provider
+  ): Promise<Transaction> {
+    // Note: should also check in caller if the approval is already given
+
+    const feeTokenAddress: string = tokenPaymasterRequest.feeQuote.tokenAddress
+    Logger.log('erc20 fee token address ', feeTokenAddress)
+
+    // TODO: For some tokens we may need to set allowance to 0 first so that would return batch of transactions
+    const requiredApproval: number = Math.floor(
```

Comment:
figured this is better to avoid running into errors. but there are bunch of calculations we can do a final review for - from types and precision point of view.